Errors initializing an object of RootObj with the C++ backend

Causes C++ compile errors

### Example
```nim
import options
type
  O = object of RootObj
   val:pointer

proc p():Option[O] = none(O)

echo p()
```

### Current Output
```
nim cpp --gc:orc -r --verbosity\:0 --hint\[Processing\]\:off --excessiveStackTrace\:on /home/deech/NimScratch/rootobj.nim
/home/deech/.cache/nim/rootobj_d/@mrootobj.nim.cpp:154:152: error: could not convert ‘{{(& NTIv2__9aZ42fy5jBLWSDBpk1jHF5g_)}, nullptr}’ from ‘<brace-enclosed initializer list>’ to ‘tyObject_O__9aZ42fy5jBLWSDBpk1jHF5g’
  154 | static NIM_CONST tyObject_Option__ecLKWfxjORDvLjP9av9ajIzQ TM__nlirijm9bpMQb6Ihp80Cbpw_21 = {{{(&NTIv2__9aZ42fy5jBLWSDBpk1jHF5g_)}, NIM_NIL}, NIM_FALSE}
      |                                                                                                                                                        ^
      |                                                                                                                                                        |
      |                                                                                                                                                        <brace-enclosed initializer list>
Error: execution of an external compiler program 'g++ -c -std=gnu++14 -funsigned-char  -w -fmax-errors=3 -fpermissive   -I/home/deech/Nim/Nim/lib -I/home/deech/NimScratch -o /home/deech/.cache/nim/rootobj_d/@mrootobj.nim.cpp.o /home/deech/.cache/nim/rootobj_d/@mrootobj.nim.cpp' failed with exit code: 1



nim-compile exited abnormally with code 1 at Wed Jun 30 20:44:15

```

### Possible Solution
Seems the reason this is happening is the generated C++ is trying to brace initialize this object:
```
struct tyObject_O__9aZ42fy5jBLWSDBpk1jHF5g : public RootObj {
void* val;
};
```
with:
```
{{(& NTIv2__9aZ42fy5jBLWSDBpk1jHF5g_)}, nullptr}
```
which is apparently illegal since `tyObject_O__9aZ42fy5jBLWSDBpk1jHF5g` [inherits from `RootObj`](https://blog.petrzemek.net/2015/10/03/structures-and-member-initializers-in-cpp/).

### Additional Information

```
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-06-30
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 36d24846b6685a44ff2185a823cdcb0f3bc236d3
active boot switches: -d:release
```

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -366,6 +366,7 @@ proc genObjectInit(p: BProc, section: TCProcSection, t: PType, a: var TLoc,
     else:
       linefmt(p, section, "$1.m_type = $2;$n", [r, genTypeInfoV1(p.module, t, a.lode.info)])
   of frEmbedded:
+    # inheritance in C++ does not allow struct initialization: bug #18836
     if not p.module.compileToCpp and optTinyRtti in p.config.globalOptions:
       var tmp: TLoc
       if mode == constructRefObj:

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -366,7 +366,7 @@ proc genObjectInit(p: BProc, section: TCProcSection, t: PType, a: var TLoc,
     else:
       linefmt(p, section, "$1.m_type = $2;$n", [r, genTypeInfoV1(p.module, t, a.lode.info)])
   of frEmbedded:
-    if optTinyRtti in p.config.globalOptions:
+    if not p.module.compileToCpp and optTinyRtti in p.config.globalOptions:
       var tmp: TLoc
       if mode == constructRefObj:
         let objType = t.skipTypes(abstractInst+{tyRef})

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -366,7 +366,7 @@ proc genObjectInit(p: BProc, section: TCProcSection, t: PType, a: var TLoc,
     else:
       linefmt(p, section, "$1.m_type = $2;$n", [r, genTypeInfoV1(p.module, t, a.lode.info)])
   of frEmbedded:
-    # inheritance in C++ does not allow struct initialization: bug #18836
+    # inheritance in C++ does not allow struct initialization: bug ##18410
     if not p.module.compileToCpp and optTinyRtti in p.config.globalOptions:
       var tmp: TLoc
       if mode == constructRefObj:

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -366,7 +366,7 @@ proc genObjectInit(p: BProc, section: TCProcSection, t: PType, a: var TLoc,
     else:
       linefmt(p, section, "$1.m_type = $2;$n", [r, genTypeInfoV1(p.module, t, a.lode.info)])
   of frEmbedded:
-    # inheritance in C++ does not allow struct initialization: bug ##18410
+    # inheritance in C++ does not allow struct initialization: bug #18410
     if not p.module.compileToCpp and optTinyRtti in p.config.globalOptions:
       var tmp: TLoc
       if mode == constructRefObj:

--- a/tests/arc/tconst_to_sink.nim
+++ b/tests/arc/tconst_to_sink.nim
@@ -1,6 +1,7 @@
 discard """
   output: '''@[(s1: "333", s2: ""), (s1: "abc", s2: "def"), (s1: "3x", s2: ""), (s1: "3x", s2: ""), (s1: "3x", s2: ""), (s1: "3x", s2: ""), (s1: "lastone", s2: "")]'''
-  cmd: "nim c --gc:arc $file"
+  matrix: "--gc:arc"
+  targets: "c cpp"
 """
 
 # bug #13240

--- a/tests/cpp/torc.nim
+++ b/tests/cpp/torc.nim
@@ -1,0 +1,15 @@
+discard """
+  targets: "cpp"
+  matrix: "--gc:orc"
+"""
+
+import std/options
+
+# bug #18410
+type
+  O = object of RootObj
+   val: pointer
+
+proc p(): Option[O] = none(O)
+
+doAssert $p() == "none(O)"

